### ISSUE TITLE
util-linux 2.39 mount: -EINVAL on overlayfs mount

### ISSUE THREAD

**Issue by fayak** [util-linux/util-linux#2287]
Hello,

On most recent version of util-linux available on Archlinux, `mount` fails to mount overlayfs in some conditions. I believe it is due to the length of argument `lowerdir`, without being able to provide further explanation. It looks like a regression from 2.38.

Demonstration:
```
$ mount -V
mount from util-linux 2.39 (libmount 2.39.0: btrfs, namespaces, idmapping, fd-based-mount, assert, debug)

$ cat test.sh 
#!/usr/bin/env bash

set -e

mkdir -p /tmp/work
cd /tmp/work

dir=aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa

mkdir -p merged
mkdir -p $dir-1 $dir-2 $dir-3 $dir-4 $dir-5

sudo mount -t overlay overlay -o lowerdir=$dir-1:$dir-2:$dir-3:$dir-4:$dir-5 merged

$ bash test.sh 
mount: /tmp/work/merged: wrong fs type, bad option, bad superblock on overlay, missing codepage or helper program, or other error.
       dmesg(1) may have more information after failed mount system call.
$ sudo pacman -U /var/backup/util-linux-2.38.1-4-x86_64.pkg.tar.zst /var/backup/util-linux-libs-2.38.1-4-x86_64.pkg.tar.zst

[ ...]

$ mount -V
mount from util-linux 2.38.1 (libmount 2.38.1: btrfs, namespaces, assert, debug)

$ bash test.sh

$ echo $?
0
```

Hope this report is appropriate

**Comment by fayak** [comment]
I believe this is hitting this limit: https://elixir.bootlin.com/linux/v6.3.5/source/fs/fsopen.c#L384
If I'm not mistaken, `lowerdir` argument is sent via `fsconfig(2)`; with a `cmd=FSCONFIG_SET_STRING`, but apparently it disallows using more than 256 chars, while `mount(2)` allows 4096

**Comment by brauner** [comment]
Yes, we already have a bug open about this. The limit for FSCONFIG_SET_STRING is fine though. See my description at:
https://lore.kernel.org/linux-unionfs/20230530-klagen-zudem-32c0908c2108@brauner/
and
https://github.com/util-linux/util-linux/issues/1992#issuecomment-1486475153

**Comment by brauner** [comment]
So this is a duplicate of https://github.com/util-linux/util-linux/issues/1992 or both.

**Comment by Andrei-Pozolotin** [comment]
more archlinux reports
https://bugs.archlinux.org/task/78702

**Comment by thesamesam** [comment]
> more archlinux reports [bugs.archlinux.org/task/78702](https://bugs.archlinux.org/task/78702)

I suggest Arch Linux stop shipping util-linux-2.39 for now then.

**Comment by karelzak** [comment]
> > more archlinux reports [bugs.archlinux.org/task/78702](https://bugs.archlinux.org/task/78702)
> 
> I suggest Arch Linux stop shipping util-linux-2.39 for now then.

It's possible to compile with --disable-libmount-mountfd-support to use mount(2) only.

**Comment by karelzak** [comment]
... and in v2.39.1 will be also possible to use LIBMOUNT_FORCE_MOUNT2=yes to stay with classic mount(2)

**Comment by thesamesam** [comment]
I'm inclined to use --default-libmount-mountfd-support for our build in Gentoo but I might expose an option for people to use the new mount API for now, off-by-default. Then in a little while (depending on how bugs go), obviously then default back to the new implementation?

Does that sound reasonable to you? Trying to balance a deluge of bug reports and also clinging to something legacy.

**Comment by karelzak** [comment]
The off-by-default is the same as releasing an RC (release candidate). Nobody will use it until someone else make it stable.

**Comment by thesamesam** [comment]
I had a feeling you might say that 😄. Okay, no problem, I'll just roll with the default (new API), and let's see what the damage is. I have a feeling the worst of it is handled already.

**Comment by karelzak** [comment]
If you want to be conservative, use --disable-libmount-mountfd-support until v2.39.2. At this time, the stuff in libmount will be stable (I hope), and we will have workarounds for kernel issues.

**Comment by karelzak** [comment]
Another possibility is to use LIBMOUNT_FORCE_MOUNT2=yes during the boot (to make sure you get a usable box) and later use the new API.

**Comment by brauner** [comment]
Groan, I picked this up porting overlayfs to the new mount api now...

**Comment by Andrei-Pozolotin** [comment]
@brauner RE:

> Groan, I picked this up

great! no, apologies, I mean: please let us know how to buy you a coffee? a beer?  :smiley: to alleviate the suffering

**Comment by t-8ch** [comment]
@brauner 
Wouldn't it still be nicer to also increase the limit for configuration options to `PAGE_SIZE` to match the legacy API?
This would avoid issues with other filesystems that users may be encountering over time.
Also it would be much easier to backport than migrations of filesystems to the new API.

Are there drawbacks for a larger limit besides that it should not be necessary in a perfect world?
(And it could be reduced again easily when all filesystems are migrated)

**Comment by brauner** [comment]
> @brauner Wouldn't it still be nicer to also increase the limit for configuration options to `PAGE_SIZE` to match the legacy API? This would avoid issues with other filesystems that users may be encountering over time. Also it would be much easier to backport than migrations of filesystems to the new API.
> 
> Are there drawbacks for a larger limit besides that it should not be necessary in a perfect world? (And it could be reduced again easily when all filesystems are migrated)

I'm not a fan. Overlayfs realistically is one of the few filesystems that runs into issue like that so I would like to avoid this until we absolutely have to. Fwiw, PAGE_SIZE is the limit for all concatenated mount options so if you have a single lowerdir parameter with PAGE_SIZE you'd not be able to mount overlayfs old or new mount api doesn't really matter.

**Comment by brauner** [comment]
The port of overlayfs to the new mount api is now in linux-next:
https://git.kernel.org/pub/scm/linux/kernel/git/next/linux-next.git/commit/?id=1784fbc2ed9c888ea4e895f30a53207ed7ee8208
https://git.kernel.org/pub/scm/linux/kernel/git/next/linux-next.git/commit/?id=b36a5780cb44a16d4384639740d87e08d4cee627
https://git.kernel.org/pub/scm/linux/kernel/git/next/linux-next.git/commit/?id=ceecc2d87f007f9fc34e847401282111c0c29786

**Comment by zeha** [comment]
https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=1784fbc2ed9c888ea4e895f30a53207ed7ee8208 is in kernel 6.5-rc1 and newer

**Comment by circum1** [comment]
FYI, stable debian with util-linux 2.41 still has this (or only related?) issue:

The following script:

```
mkdir -p /tmp/overlay-bugtest/{lower,merged,work,upper}
for n in 1 2; do
  mkdir -p "/tmp/overlay-bugtest/lower/00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000-${n}"
done

sudo mount -t overlay overlay -o lowerdir=/tmp/overlay-bugtest/lower/00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000-1:/tmp/overlay-bugtest/lower/00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000-2,upperdir=/tmp/overlay-bugtest/upper,workdir=/tmp/overlay-bugtest/work /tmp/overlay-bugtest/merged
```

Results in the error message
```

mount: /tmp/overlay-bugtest/merged: wrong fs type, bad option, bad superblock on overlay, missing codepage or helper program, or other error.
       dmesg(1) may have more information after failed mount system call.
```

Using busybox mount, or shortening the lowerdir directory names makes the error disappear.

Kernel version is 6.7.12.

**Comment by t-lo** [comment]
Confirming on Fedora 43 / util-linux 2.41.1 with ` fd-based-mount` feature enabled.

`overlayfs` options string is limited to 256 characters while "classic" mount is limited to `PAGE_SIZE` (usually 4096). While this is true for all filesystems supported by `mount(8)`, overlayfs is particularly impacted as mount option strings tend to be longer than usual - especially with many lowerdirs and/or long lowerdir directory names.

Repro
```bash
dirs=$(echo /tmp/o/{1..300})
mkdir -p $dirs /tmp/o/m
mount -t overlay -o lowerdir=${dirs// /:} ovtest /tmp/o/m # fails
busybox mount -t overlay -o lowerdir=${dirs// /:} ovtest /tmp/o/m # succeeds
```
